Add DnD tests issue #3182

aloowood opened the pull request.

merks commented on features/.project:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>features</name>
+	<comment></comment>
+	<projects>
+	</projects>
+	<buildSpec>
+	</buildSpec>
+	<natures>
+	</natures>
+</projectDescription>
```
Adding a .project file here is not needed and not desired.

merks commented on tests/org.eclipse.ui.tests.dnd/src/org/eclipse/ui/tests/dnd/DetachedDropTarget.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
```
I guess you copied this from somewhere.

merks commented on tests/org.eclipse.ui.tests.dnd/src/org/eclipse/ui/tests/dnd/DetachedWindowDragTest.java:
```diff
@@ -0,0 +1,26 @@
+package org.eclipse.ui.tests.dnd;
```
Every java file needs a "standard" copyright header.

merks commented on tests/org.eclipse.ui.tests.dnd/src/org/eclipse/ui/tests/dnd/DragTest.java:
```diff
@@ -0,0 +1,95 @@
+
+    IEditorPart editor1, editor2, editor3;
+
+    static WorkbenchWindow window;
```
Can this be an IWorkbenchWindow?  Should any of these things really be static?

merks commented on tests/org.eclipse.ui.tests.dnd/src/org/eclipse/ui/tests/dnd/DragTest.java:
```diff
@@ -0,0 +1,95 @@
+    IEditorPart editor1, editor2, editor3;
+
+    static WorkbenchWindow window;
+    static WorkbenchPage page;
```
Can this be an IWorkbenchPage?

merks commented on tests/org.eclipse.ui.tests.dnd/src/org/eclipse/ui/tests/dnd/DragTestSuite.java:
```diff
@@ -0,0 +1,207 @@
+				continue;
+			}
+
+			DragTest newTest = new DragTest(dragSource, dropTarget);
```
Avoid warnings.

merks reviewed (changes requested):
Hopefully it's possible to use IWorkbenchWindow and IWorkbenchPage instead of non-API implementation classes.

In general, please keep the PR to a single commit.  Using amend and force push for that purpose rather than keep adding commits.

aloowood commented:
I have applied the requested changes, cleaned up all warnings, and rebased on top of the latest master. Ready for another review.

HeikoKlare commented on tests/org.eclipse.ui.tests.dnd/src/org/eclipse/ui/tests/dnd/DragOperations.java:
```diff
@@ -0,0 +1,92 @@
+	 * This method should eventually replace the original one once the Workbench has been updated
+	 * to handle Views and Editors without distincton.
+	 */
+	public static void drag(IWorkbenchPart part, TestDropLocation target, boolean wholeFolder) {
```
I don't think the tests can really work of this essential method still fails, do they?

HeikoKlare commented on tests/org.eclipse.ui.tests.dnd/src/org/eclipse/ui/tests/dnd/DragTestSuite.java:
```diff
@@ -0,0 +1,218 @@
+import junit.framework.Test;
+
+@RunWith(AllTests.class)
+@Ignore("DND support not implemented yet in E4")
```
You need to remove the `@Ignore` and should set the test suite as the class executed for the bundle in the `build.properties` like in the other test bundles, e.g.:
https://github.com/eclipse-platform/eclipse.platform.ui/blob/223dfe7383547fdead8d59f43fa7d8ad6b2e2022/tests/org.eclipse.ui.tests/build.properties#L32

We finally also need to add a proper `test.xml` to have the tests executed in Jenkins.

HeikoKlare commented on tests/org.eclipse.ui.tests.dnd/src/org/eclipse/ui/tests/dnd/DragTestSuite.java:
```diff
@@ -0,0 +1,218 @@
+
+@RunWith(AllTests.class)
+@Ignore("DND support not implemented yet in E4")
+public class DragTestSuite {
```
This pretty much looks like a JUnit-3-style TestSuite. Since we are about to migrate everything to JUnit 5 (I just migrated everything to JUnit 4), reintroducing these tests should also incorporate and up-to-date style of the test suite.

HeikoKlare reviewed (commented):
Thank you for the update. Before we can take a deeper look into the proposal, there are some general concerns to address.

aloowood committed:
Migrate Drag-and-Drop tests to JUnit 5 and provide basic DragOperations implementation

aloowood commented:
I incorporated the feedback, rebased onto the current master, and squashed everything into a single commit. Thanks in advance for taking another look.

HeikoKlare reviewed (commented):
Thank you for the latest update. Unfortunately, there is still no proper test suite and no test executed at all.
The `DragTestSuite` class is not a proper test suite (neither JUnit 4 nor JUnit 5). It is just a plain JUnit 5 test class with a paramaterized test method that effectively does nothing but instantiating the `DragTest` class without executing anything at all.
You will also notice that when executing the tests each of them is executed in less than millisecond (at least on my system) which is unlikely for a UI test.

HeikoKlare commented:
One additional hint to ensure that you first come to a functional state of the tests before pinging us for review: make sure that the tests that you add/update are actually executed by testing it locally. Run the tests in your IDE and in case you are not familiar with how to set up JUnit 4/5 tests / test suites and are in doubt if it works, you can e.g. set a breakpoint in the actual test methods to ensure that they are really executed. You should also run the single test classes individually to see that they also work without running the complete test suite.

aloowood committed:
Add clean DnD Test

aloowood commented:
Hi @HeikoKlare,
I’m working on the DnD tests and wanted to ask whether I should use SWTBot or the PDE test framework for this. I noticed that SWTBot isn’t fully compatible with Eclipse 4.37, so I’d like to confirm which approach you’d recommend.

HeikoKlare commented:
What do you mean with the PDE test framework? I don't think we currently have either a dependency to SWTBot or any PDE artifacts in Platform UI. Can you explain why you need either of those?

aloowood commented:
I'm trying to write DnD tests, but I'm not entirely sure which approach is preferred in Platform UI. What method would you recommend for this? I want to be sure I'm following the correct approach.
